Refactor code and Fix specs

kushal-dreamspring opened the pull request:
- Update imports of models and controllers
- Refactor unpark.spec.js
- Remove cookieParser library
- Refactor file read and write
- Refactor isCarParked and isLotFull into seperate functions
- Set null as default timestamp
- Add parking lot model
- Add specs for car, slots and parking lot models
- Move controller specs into one file in specs/controllers folder
- Rename ParkingLotController.js to controller.js
- Add regex based browser side validation to registration number
- Add custom validation message on browser side validation
- Refactor browser-side js scripts
- Replace var with const throughout the project
- Refactor regex for registration number
- Use camelCase instead of snake_case
- Use better variable names for specs
- Rename findEmptySlot to emptySlotIndex
- Replace literal 10 with DEFAULT_LOT_SIZE declared in config/constants.js
- Remove unused code

kushal-dreamspring committed:
Fix Specs

- Update imports of models and controllers
- Refactor unpark.spec.js

kushal-dreamspring committed:
Refactor code

- Remove cookieParser
- Refactor file read and write
- Refactor isCarParked and isLotFull into seperate functions
- Set null as default timestamp

kushal-dreamspring committed:
Code Refactoring

- Add parking lot model
- Add specs for car, slots and parking lot models
- Move controller specs to specs/controllers folder
- Rename ParkingLotController.js to controller.js
- Add regex based browser side validation to registration number
- Add custom validation message on browser side validation
- Refactor browser-side js scripts
- Replace var with const throughout the project
- Refactor regex for registration number

kushal-dreamspring committed:
Code Refactoring

- Add specs for car, slots and parking lot models
- Move controller specs to specs/controllers folder

kushal-dreamspring committed:
Move all controller related specs into one file

jinifrancis commented on spec/controllers/controllers.spec.js:
```diff
@@ -0,0 +1,209 @@
+const Slot = require("../../models/slot");
+
+describe("Initialize suite", function () {
+  beforeAll(helper.deleteDB);
```
There is just one spec in this section - so there is no need for `beforeAll`. The `deleteDB` can go inside the spec itself.

jinifrancis commented on spec/controllers/controllers.spec.js:
```diff
@@ -0,0 +1,209 @@
+    }
+  });
+
+  // afterEach(helper.deleteDB);
```
Remove line if not required

jinifrancis commented on spec/controllers/controllers.spec.js:
```diff
@@ -0,0 +1,209 @@
+
+  it("should park a car with valid registration number", async function () {
+    try {
+      const registration_number = "UP32EA7196";
```
camelCase variable name

jinifrancis commented on spec/controllers/controllers.spec.js:
```diff
@@ -0,0 +1,209 @@
+        );
+
+      const data = JSON.stringify(await controller.getAllCars());
+      const recentCars = {
```
Incorrect variable name for All cars suite test variable

jinifrancis replied:
The feedback item is pending

jinifrancis commented on controllers/controller.js:
```diff
@@ -0,0 +1,58 @@
+    if (lot.findCarSlot(registration_number) !== -1)
+      return { error: "Vehicle already parked" };
+
+    let index = lot.findEmptySlot();
```
Rename to `emptySlotNumber` or `emptySlotLocation` to indicate what the variable name holds

jinifrancis commented on controllers/controller.js:
Camelcase variable name for the registration number, slot number etc

jinifrancis commented on public/javascripts/script.js:
```diff
@@ -95,48 +95,35 @@ function get_recent_cars() {
-          timeStyle: "medium",
+            timeZone,
+            dateStyle: "medium",
+            timeStyle: "medium",
```
Remove tab spaces

jinifrancis commented on spec/helper/db.helper.js:
```diff
@@ -1,12 +1,12 @@
 exports.newDB = async () => {
   try {
-    const lot = [...Array(10).keys()].map((i) => new ParkingSlot(i));
+    const lot = [...Array(10).keys()].map((i) => new Slot(i));
```
Move the size of the Array to a constant which can be used in all places, instead of repetitive usage of the magic number 10 in every location. This will help reduce the number to changes required if we need to increase or decrease the parking lot size at any point of time.

jinifrancis replied:
The feedback item is pending

jinifrancis replied:
I still see that we are using the number `10` here and in a lot of other places instead of the constant. Did you work ok moving the parking lot size to a constant? Reach out to me if you have questions regarding this review point.

> 

> Move the size of the Array to a constant which can be used in all places, instead of repetitive usage of the magic number 10 in every location. This will help reduce the number of changes required if we need to increase or decrease the parking lot size at any point in time.

jinifrancis commented on spec/models/car.spec.js:
```diff
@@ -0,0 +1,16 @@
+
+  it("should return registration number of car", function () {
+    const registration_number = "UP32EA7196";
+    const c = new Car(registration_number);
```
rename `c` to `car`

jinifrancis commented on spec/models/parkingLot.spec.js:
```diff
@@ -0,0 +1,81 @@
+  let lot = null;
+  beforeEach(function () {
+    lot = new ParkingLot();
+  });
```
Add new line after this block

jinifrancis commented on spec/models/parkingLot.spec.js:
```diff
@@ -0,0 +1,81 @@
+    lot = new ParkingLot();
+  });
+  it("should create a list of 10 empty slots", function () {
+    expect(lot.lot.length).toEqual(10);
```
Move number of empty slots to a constant as mentioned earlier and use it here

kushal-dreamspring committed:
Fix bugs

- Use camelCase instead of snake_case
- Use variable for slot size
- Use better variable names for specs
- Remove unused code

kushal-dreamspring committed:
Rename findEmptySlot to emptySlotIndex

jinifrancis commented on models/car.js:
```diff
@@ -1,11 +1,11 @@
-  constructor(registration_number) {
-    this.registration_number = registration_number;
+  constructor(registrationNumber) {
+    this.registration_number = registrationNumber;
```
Make sure variable names are camelCase

jinifrancis commented on models/parkingLot.js:
```diff
@@ -0,0 +1,76 @@
+
+  insertData(data) {
+    this.lot = data.map((slot) => {
+      const object = new Slot(slot.slot_no);
```
Missing camelCase for `slot_no`

jinifrancis commented on models/parkingLot.js:
```diff
@@ -0,0 +1,76 @@
+      const object = new Slot(slot.slot_no);
+
+      if (slot.car)
+        object.parkCar(slot.car.registration_number, slot.timestamp);
```
Missing camelCase for registration_number

jinifrancis reviewed (changes requested):
I am still seeing a lot of places where variable names or object keys are not in camelCase as per JS standards. Please make necessary changes

kushal-dreamspring committed:
Change all variables and data to camelCase

jinifrancis commented on models/parkingLot.js:
```diff
@@ -0,0 +1,76 @@
+const Slot = require("./slot");
+
+class ParkingLot {
+  constructor(fileData, lotSize = 10) {
```
Why is the default value for `lotSize` mentioned here directly as 10? Wasn't it moved to a constant?

jinifrancis commented on spec/controllers/controllers.spec.js:
```diff
@@ -0,0 +1,208 @@
+      await controller.parkCar(registrationNumber, timestamp);
+
+      const data = await fs.readFile(path.join("db", "db.json"), "utf8");
+      const lot = [...Array(10).keys()].map((i) => new Slot(i));
```
Use constant instead of `10`

jinifrancis commented on spec/controllers/controllers.spec.js:
```diff
@@ -0,0 +1,208 @@
+    }
+  });
+
+  it("should not park more than 10 cars", async function () {
```
Try to think of this spec from the perspective of - "should not park more than the size of the parking lot" and use constant for parking lot size instead of using the number `10`

kushal-dreamspring committed:
Replace literal 10 with constants

jinifrancis reviewed (changes requested):
Can you think of a way to declare the lot size only once, and use the same constant at all places instead of initializing a variable wherever it is needed?

kushal-dreamspring committed:
Refactor Code

- Declare DEFAULT_LOT_SIZE in spec helper

kushal-dreamspring committed:
Move DEFAULT_LOT_SIZE to config/constants.js

- Add config/constants.js
- Import DEFAULT_LOT_SIZE from config/constants.js in ParkingLot Model
- Import DEFAULT_LOT_SIZE from config/constants.js in Specs
- Remove DEFAULT_LOT_SIZE declaration from ParkingLot Model & db helper

jinifrancis reviewed (approved):
This PR looks good to be merged 🎉
